### DISCUSSION TITLE
Algorithm for creation of block-cut tree

### DISCUSSION THREAD

**Discussion by danielolsen** [networkx/networkx#5395] [Algorithms]
This may already be a part of NetworkX, but if so, I wasn't able to find it by looking through the documentation.

I want to be able to create a block-cut graph representation of a source graph. Is this currently possible?

**Comment by dschult** [comment]
One of the troubles with any field, but graph theory in particular is that there are often multiple names for the same thing.  Can you describe what you mean by "block-cut graph representation"?  We don't have anything called that (that I recall), but maybe we have it called something else... :}

**Reply by danielolsen** [reply]
My graph theory background is rudimentary at best, I'm pulling the terminology from the wikipedia page: https://en.wikipedia.org/wiki/Biconnected_component#Block-cut_tree. It seems that this could be generated by calling `biconnected_components` and `articulation_points` and stitching the outputs of each back together, but I would guess that there is some duplicated computation using that path.

**Comment by danielolsen** [comment]
Digging a little deeper, it seems that both the biconnected components and the articulation points are generated by the same `_biconnected_dfs` private function. I wonder if instead of `components=True` or `components=False`, this function could be refactored to yield either the components, the articulation points, or both. If we get both, then I think that would help us build a block-cut graph more efficiently than by using repeated calls to `_biconnected_dfs`. This would require some refactor of the public functions within `networkx.algorithms.components.biconnected` though, plus a new public function to re-stitch the block-cut graph from a generator returning both.

**Comment by dschult** [comment]
I believe your summary is correct. We don't currently have a function to compute the block-cut tree and we have the machinery to do it.  You would need to change the private function to allow both components and articulation points. I'm not sure if it is most effective to provide a third option for the existing `components` parameter, or to e.g. make `_biconnected_dfs` yield them both in a predictable way that allows the public functions to ignore the items it doesn't want.  If it takes extra work to find both then we shouldn't do that. And if the pattern is tricky then we also shouldn't do that. But if they would always alternate between component and articulation point, then that might be a nice interface.  In any case, you would indeed need to adapt the public functions to work with the changes in the private function and you'd want a public function to return the block-cut tree itself. 
Thanks for this summary. We'd be interested if you are up for putting this together.
